### PR TITLE
Refine gradient accumulation (on device training)

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: Refine gradient accumulation （on device training）

Few major refinements:
1. Disable accumulate_gradient_ for Module initially, otherwise, accumulation will be done in the first step. 
2. Perf improvement around inplace accumulator. 

   Different data scale, we see different perf diff when we compare master with the initial commits ([061232b](https://github.com/microsoft/onnxruntime/pull/12363/commits/061232b20ab429b28e43abaecbc0e094aafe7122)) in this PR.
  image
  
    Based on this results, a branch is introduced for making both small data or big data happy (e.g. faster).

    
    image

     Noted: some of the small data show degradation, which should be normal variances. 

4. Add more synthetic inputs. 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu [hunk 1]
```diff
@@ -26,26 +49,38 @@ void InPlaceAccumulatorImpl(
...
+#else
+  const int num_elements_per_thread = GridDim::maxElementsPerThread;
+  const int num_threads_per_block = GridDim::maxThreadsPerBlock;
+#endif
+
+  int blocksPerGrid = static_cast<int>(CeilDiv(count, num_threads_per_block * num_elements_per_thread));
   CUDA_LONG N = static_cast<CUDA_LONG>(count);
-  _InPlaceAccumulator<T, T_GRAD><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(
+
+  _InPlaceAccumulator<T, T_GRAD, num_threads_per_block, num_elements_per_thread><<<blocksPerGrid, num_threads_per_block, 0, stream>>>(
```
Comment (baijumeswani) [inline orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu:66]: please fix cpplint error. (line > 120 chars)

File: orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu [hunk 2]
```diff
@@ -26,26 +49,38 @@ void InPlaceAccumulatorImpl(
...
+
+#ifdef USE_ROCM
+  const int num_elements_per_thread = 2;
+  const int num_threads_per_block = 512;
+#else
+  const int num_elements_per_thread = GridDim::maxElementsPerThread;
+  const int num_threads_per_block = GridDim::maxThreadsPerBlock;
+#endif
+
+  int blocksPerGrid = static_cast<int>(CeilDiv(count, num_threads_per_block * num_elements_per_thread));
```
Comment (baijumeswani) [inline orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu:63]: `const int`

File: orttraining/orttraining/training_api/include/module.h [hunk 3]
```diff
@@ -100,7 +100,7 @@ struct Module {
   std::vector<std::string> eval_output_names_;
   std::vector<OrtValue> weights_;
   std::vector<OrtValue> gradients_;
-  bool accumulate_gradient_ = true;
+  bool accumulate_gradient_ = false;
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/training_api/include/module.h:103]: as we talked offline, this can avoid the first step do accumulation, although this being false or true, result is same.

File: orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu [hunk 4]
```diff
@@ -26,26 +58,47 @@ void InPlaceAccumulatorImpl(
...
   CUDA_LONG N = static_cast<CUDA_LONG>(count);
-  _InPlaceAccumulator<T, T_GRAD><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(
-      gradient_buffer,
-      gradient,
-      accumulated_gradient,
-      N);
+
+  // This number is tuned among many data scales that can fit one GPU card.
+  // The data scales are also listed in unit tests.
+  const size_t large_data_scale_threshold = 4096 * 768;
```
Comment (baijumeswani) [inline orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu:69]: How do we come up with this figure? Is it empirical? Can we put in an unnamed namespace as a constexpr?

**Review by baijumeswani — APPROVED** [review]

File: orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu (hunk 4 above)
Comment (pengwa) [reply orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu:69]: As the comment said, "This number is tuned among many data scales that can fit one GPU card. The data scales are also listed in unit tests." I profile a set of input sizes, and the perf result  is shown in the PR desc, and pick up this one.

Comment (pengwa) [reply orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.cu:69]: Fixed the constexpr.

**Review by baijumeswani — APPROVED** [review]
